guppy/src/graph/cargo/cargo_api.rs

```diff
@@ -254,9 +254,30 @@ impl<'g> CargoSet<'g> {
         initials: FeatureSet<'g>,
         features_only: FeatureSet<'g>,
         opts: &CargoOptions<'_>,
+    ) -> Result<Self, Error> {
+        Self::new_internal(initials, features_only, None, opts)
+    }
+
+    /// Like `Cargo.new`, but allows inspecting and influencing dependency traversal with a custom
+    /// `resolver`.
+    pub fn with_resolver(
+        initials: FeatureSet<'g>,
+        features_only: FeatureSet<'g>,
+        mut resolver: impl PackageResolver<'g>,
+        opts: &CargoOptions<'_>,
+    ) -> Result<Self, Error> {
+        Self::new_internal(initials, features_only, Some(&mut resolver), opts)
+    }
```

Comment:
Regarding testing: we currently have comparison tests between guppy and cargo, but those only work for capabilities supported by both (package link filtering is not supported by Cargo). The hakari and summary fixture tests also ensure that omitted packages work.

For package link filtering, I think it would be useful to have a test for this. We have a somewhat complex workspace that's checked in as a fixture: https://github.com/guppy-rs/guppy/tree/main/fixtures/workspace/inside-outside -- I'd add tests to ensure that all of the scenarios are covered. (For example, the current code does not appear to do proc macro redirect filtering, so that should be fixed and a test added for it.)